### PR TITLE
Locally execute subqueries that use only subplan results

### PR DESCRIPTION
**Pull request by hanefi**
DESCRIPTION: Local execution of subqueries that use only intermediate results

If there is a subquery or CTE that uses only the results of a subplan we used to push intermediate results to a random worker using round robin policy, and execute it there. This PR enables local execution of those queries.

We introduce a new GUC for this functionality `citus.enable_local_execution_planning` that is enabled by default. If this flag is set, we do not push the intermediate results to a worker to execute a subquery, and execute it locally instead. This behaviour is independent of the `citus.enable_local_execution`.

Things to test:
- [x] There exists an entry for the coordinator in  `pg_dist_node`
- [x] There is not an entry for the coordinator in  `pg_dist_node`
- [x] MX nodes
- [x] append distributed tables
- [x] `INSERT .. SELECT ..` queries

Regression tests:

|  	| local_execution=ON 	| local_execution=OFF 	|
|------------------------------	|--------------------	|---------------------	|
| local_execution_planning=ON 	|✅ intermediate_result_pruning  	| manually tested 	|
| local_execution_planning=OFF 	|✅ failure_cte_subquery	| manually tested	|

### CODE REVIEW

File: src/test/regress/expected/multi_task_assignment_policy.out [hunk 1]
```diff
@@ -320,7 +320,7 @@ $cmd$, 'task_assignment_test_table_2');
 SELECT count(DISTINCT value) FROM explain_outputs;
  count
 ---------------------------------------------------------------------
-     2
```
Comment (hanefi) [inline src/test/regress/expected/multi_task_assignment_policy.out]: I need to update the comments in this test.

I am planning to add some more test cases here to make sure we still use round-robin assignments when the flag is set, and locally execute the query otherwise

File: src/backend/distributed/executor/local_executor.c [hunk 2]
```diff
@@ -93,9 +93,10 @@
 
 /* controlled via a GUC */
 bool EnableLocalExecution = true;
+bool EnableLocalExecutionPlanning = true;
 bool LogLocalCommands = false;
 
-bool LocalExecutionHappened = false;
+bool LocalPlacementExecutionHappened = false;
```
Comment (marcocitus) [inline src/backend/distributed/executor/local_executor.c]: Maybe `TransactionAccessedLocalPlacement`

**Comment by marcocitus** [comment]
It seems we're currently connecting to localhost. Can we use local execution?

```sql
postgres=# WITH a AS (SELECT * FROM test) SELECT * FROM a;
...
LOG:  issuing SELECT x, y FROM test_102039 test WHERE true
DETAIL:  on server marco@localhost:9701 connectionId: 2
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 4, '2020-01-14 00:07:25.469807+01');
DETAIL:  on server marco@localhost:5432 connectionId: 17
LOG:  issuing SELECT x, y FROM (SELECT intermediate_result.x, intermediate_result.y FROM read_intermediate_result('8_1'::text, 'binary'::citus_copy_format) intermediate_result(x integer, y integer)) a
DETAIL:  on server marco@localhost:5432 connectionId: 17
```

Would expect it to say:
```sql
LOG:  executing the command locally: SELECT x, y FROM (SELECT intermediate_result.x, intermediate_result.y FROM read_intermediate_result('8_1'::text, 'binary'::citus_copy_format) intermediate_result(x integer, y integer)) a
```

File: src/include/distributed/subplan_execution.h [hunk 3]
```diff
@@ -32,6 +32,7 @@ extern void ExecuteSubPlans(DistributedPlan *distributedPlan);
 typedef struct IntermediateResultsHashEntry
 {
 	char key[NAMEDATALEN];
+	bool writeLocalFile;
```
Comment (pykello) [inline src/include/distributed/subplan_execution.h]: update the comments before the struct definition to describe what this is used for.

File: src/backend/distributed/planner/intermediate_result_pruning.c [hunk 4]
```diff
@@ -206,17 +220,23 @@ MakeIntermediateResultHTAB()
...
 	foreach(nodeIdCell, entry->nodeIdList)
 	{
-		WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+		uint32 nodeId = lfirst_int(nodeIdCell);
+		WorkerNode *workerNode = LookupNodeByNodeId(nodeId);
+
+		if (workerNode == NULL)
+		{
+			elog(DEBUG4, "Failed Node Lookup with Id %d, write to local file", nodeId);
+			entry->writeLocalFile = true;
```
Comment (pykello) [inline src/backend/distributed/planner/intermediate_result_pruning.c]: how can this happen?

File: src/backend/distributed/planner/intermediate_result_pruning.c [hunk 5]
```diff
@@ -228,6 +248,22 @@ FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
...
+	if (entry->writeLocalFile)
+	{
+		WorkerNode *workerNode = NULL;
+		int32 localGroupId = GetLocalGroupId();
+
+		foreach_ptr(workerNode, workerNodeList)
+		{
+			if (workerNode->groupId == localGroupId)
+			{
+				workerNodeList = list_delete(workerNodeList, workerNode);
```
Comment (pykello) [inline src/backend/distributed/planner/intermediate_result_pruning.c]: `list_delete` is for lists of nodes. We should use `list_delete_ptr()` here, as `WorkerNode` is not a postgres node structure.

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 4 above)
Comment (hanefi) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: If we have a dummy placement that points to the coordinator that does not have a valid record in `pg_dist_node` I assumed that it should be written to a local file.

I will introduce a new placeholder such as `INVALID_NODE_ID=0` that will be used for this purpose. Does this sound good?

File: src/test/regress/expected/intermediate_result_pruning.out [hunk 6]
```diff
@@ -4,6 +4,8 @@ SET citus.log_intermediate_results TO TRUE;
 SET citus.shard_count TO 4;
 SET citus.next_shard_id TO 1480000;
 SET citus.shard_replication_factor = 1;
+SET citus.enable_local_execution_planning TO 1;
+SET citus.enable_local_execution TO 1;
```
Comment (marcocitus) [inline src/test/regress/expected/intermediate_result_pruning.out]: Can you add some tests to confirm using citus.log_remote_commands to confirm that local execution is being used?

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 4 above)
Comment (hanefi) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: When creating a dummy placement, the `nodeId` field will now be `DUMMY_NODE_ID = -1` and before doing a lookup, I will check if this instance is indeed a dummy placement.

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 5 above)
Comment (hanefi) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: thanks for pointing this out.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 7]
```diff
@@ -2219,6 +2207,51 @@ PlanRouterQuery(Query *originalQuery,
...
+ * Otherwise an active worker is selected using round robin policy to create the
+ * dummy placement. In case there are no active workers, NULL pointer is returned.
+ */
+static ShardPlacement *
+CreateDummyPlacement(void)
+{
+	static uint32 zeroShardQueryRoundRobin = 0;
+	ShardPlacement *dummyPlacement = CitusMakeNode(ShardPlacement);
+
+	if (!EnableLocalExecutionPlanning)
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: Perhaps we should use the citus.task_assignment_policy here? i.e. if it's round-robin we pick a random node and otherwise we pick localhost.

Comment (hanefi) [reply src/backend/distributed/planner/multi_router_planner.c]: Does this mean we should remove `citus.enable_local_execution_planning` completely? @onderkalaci suggested I create a new flag that is enabled by default. I guess this GUC is no longer needed if I use `TaskAssignmentPolicy`


```c
if(TaskAssignmentPolicy == TASK_ASSIGNMENT_ROUND_ROBIN)
{ /* round-robin */ }
else
{ /* dummy placement pointing at local node */ }
```

File: src/backend/distributed/connection/placement_connection.c [hunk 8]
```diff
@@ -1164,16 +1164,36 @@ ColocatedPlacementsHashCompare(const void *a, const void *b, Size keysize)
...
- * distirbuted transaction and already executed at least one command that
- * accessed to a placement.
+ * AnyConnectionModifiedPlacements checks the entries in ConnectionPlacementHash
+ * and determines if we executed a command that modified a placement. This is
+ * useful to detect whether we're in a distributed transaction and already
+ * executed at least one command that modified a placement.
  */
 bool
-AnyConnectionAccessedPlacements(void)
+AnyConnectionModifiedPlacements(void)
```
Comment (onderkalaci) [inline src/backend/distributed/connection/placement_connection.c]: Do we need this change for the PR?

We've not done this in the past because of performance reasons. You have to iterate over the connection hash now.

https://github.com/citusdata/citus/pull/2938#pullrequestreview-285478511

I'd not do it in this PR if it is not strictly necessary

File: src/backend/distributed/planner/intermediate_result_pruning.c [hunk 9]
```diff
@@ -101,11 +98,13 @@ RecordSubplanExecutionsOnNodes(HTAB *intermediateResultsHash,
 		IntermediateResultsHashEntry *entry = SearchIntermediateResult(
 			intermediateResultsHash, resultId);
 
-		/* no need to traverse the whole plan if all the workers are hit */
-		if (list_length(entry->nodeIdList) == workerNodeCount)
+		/*
+		 * There is no need to traverse the whole plan if the intermediate result
```
Comment (onderkalaci) [inline src/backend/distributed/planner/intermediate_result_pruning.c]: I get this log:
```
DEBUG:  Subplan 61_1 will be sent to localhost:9700
DEBUG:  Subplan 61_1 will be sent to localhost:9701
DEBUG:  Subplan 61_1 will be sent to localhost:9702
DEBUG:  Subplan 61_1 will be written to local file
DEBUG:  Subplan 61_2 will be written to local file
```

I'm operation on `localhost:9700`, is this expected to see both written to local file and `will be sent to localhost:9700`?

Also, I already have 3 nodes only, why don't we see `Subplan %s is used in all workers`? 

I'm asking whether are these logging issues or we're actually sending the file twice?

the query that I use on MX worker is:

```SQL
WITH foo AS (SELECT * FROM users_table OFFSET 5) SELECT user_id, count(*) FROM foo JOIN users_table USING (user_id) GROUP BY user_id HAVING (max(foo.value_1) > (SELECT max(user_id) FROM foo)) ORDER BY 1 DESC LIMIT 5;
```

File: src/backend/distributed/connection/placement_connection.c [hunk 10]
```diff
@@ -1164,16 +1164,36 @@ ColocatedPlacementsHashCompare(const void *a, const void *b, Size keysize)
...
+	{
+		/* check if we have opened a connection and modified this placement */
+		if (entry->primaryConnection != NULL &&
+			(entry->primaryConnection->hadDDL || entry->primaryConnection->hadDML))
+		{
+			/* abandon sequential search before completion */
+			hash_seq_term(&status);
+			return true;
+		}
+		entry = (ConnectionPlacementHashEntry *) hash_seq_search(&status);
```
Comment (marcocitus) [inline src/backend/distributed/connection/placement_connection.c]: Onder reminded me that I previously disapproved of this approach:
https://github.com/citusdata/citus/pull/2938#discussion_r322220263

Is there a faster way to achieve this?

File: src/backend/distributed/planner/intermediate_result_pruning.c [hunk 11]
```diff
@@ -228,6 +252,22 @@ FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
...
+	if (entry->writeLocalFile)
+	{
+		WorkerNode *workerNode = NULL;
+		int32 localGroupId = GetLocalGroupId();
+
+		foreach_ptr(workerNode, workerNodeList)
+		{
+			if (workerNode->groupId == localGroupId)
+			{
+				workerNodeList = list_delete_ptr(workerNodeList, workerNode);
```
Comment (marcocitus) [inline src/backend/distributed/planner/intermediate_result_pruning.c]: Does this cause any issues given that we're iterating through this list? (not sure)

File: src/backend/distributed/connection/placement_connection.c (hunk 8 above)
Comment (marcocitus) [reply src/backend/distributed/connection/placement_connection.c]: I'm guessing we need to change something, since one of the main use cases is:
`WITH a AS (SELECT ... FROM distributed_table ...) SELECT * FROM a ...;` which clearly accesses placements.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 7 above)
Comment (marcocitus) [reply src/backend/distributed/planner/multi_router_planner.c]: I'd suggest removing the GUC.

File: src/backend/distributed/connection/placement_connection.c (hunk 10 above)
Comment (hanefi) [reply src/backend/distributed/connection/placement_connection.c]: I am looking for another way to achieve this. No luck so far

File: src/test/regress/expected/failure_cte_subquery.out [hunk 12]
```diff
@@ -53,6 +54,43 @@ ERROR:  server closed the connection unexpectedly
...
+     	users_table, events_table
+     WHERE
+     	users_table.user_id = events_table.user_id AND
+     event_type IN (1,2,3,4)
+     ORDER BY 1 DESC LIMIT 5
+     ) as foo
+	  WHERE foo.user_id = cte.user_id;
+ count
+---------------------------------------------------------------------
+     0
```
[the remark this answers is not in the document]
Comment (hanefi) [reply src/test/regress/expected/failure_cte_subquery.out]: inserted some rows, now the count is 8

File: src/backend/distributed/planner/multi_router_planner.c (hunk 7 above)
Comment (hanefi) [reply src/backend/distributed/planner/multi_router_planner.c]: I removed `enable_local_execution_planning` and planner will prefer local execution if `TASK_ASSIGNMENT_ROUND_ROBIN` is not in effect.

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 11 above)
Comment (hanefi) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: I checked `list_delete_ptr()` implementation, and there is no harm in using it this way. The only downside is `list_delete_ptr()` iterates through the linked list to update the link on previous cell.

https://github.com/postgres/postgres/blob/master/src/backend/nodes/list.c#L795-L812

File: src/backend/distributed/executor/adaptive_executor.c [hunk 13]
```diff
@@ -163,129 +165,6 @@
 #include "utils/timestamp.h"
 
 
-/*
- * DistributedExecution represents the execution of a distributed query
- * plan.
- */
-typedef struct DistributedExecution
```
Comment (onderkalaci) [inline src/backend/distributed/executor/adaptive_executor.c]: we'd prefer to keep this struct private, can we do it another way?

File: src/backend/distributed/executor/local_executor.c [hunk 14]
```diff
@@ -405,10 +406,11 @@ ShouldExecuteTasksLocally(List *taskList)
...
 		 *
 		 * Note that we shouldn't use local execution if any distributed execution
-		 * has happened because that'd break transaction visibility rules and
-		 * many other things.
+		 * modified a distributed table because that'd break transaction visibility
+		 * rules and many other things. Therefore it is the callers responsibility
+		 * to make sure no modifications are made in a distributed transaction
 		 */
-		return !AnyConnectionAccessedPlacements();
+		return !TransactionModifiedDistributedTable(execution);
```
Comment (marcocitus) [inline src/backend/distributed/executor/local_executor.c]: Would it be sufficient to just check `XactModificationLevel == XACT_MODIFICATION_DATA`?

Comment (onderkalaci) [reply src/backend/distributed/executor/local_executor.c]: I think no because that flag is also set after local execution modified any placement. The basic idea of that check is to make sure that no remote access happened before, excluding local accesses

Comment (marcocitus) [reply src/backend/distributed/executor/local_executor.c]: Stepping back. I think the main thing we wanted to achieve here is to not use local execution if we already connected to localhost (e.g. to do DDL). Should we add a `ConnectedToLocalhost` flag that becomes true if we're opening a connection to groupId == GetLocalGroupId()? Eventually this flag will disappear once we support local execution for everything.

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 9 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: I fixed this

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 11 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: Well, even if it doesn't cause any issues, it looks weird. Instead, I got the copy of a shallow copy of the workerNodeList

File: src/test/regress/expected/intermediate_result_pruning.out (hunk 6 above)
Comment (onderkalaci) [reply src/test/regress/expected/intermediate_result_pruning.out]: makes sense, we've another GUC, which logs local execution even on `DEBUG1`, `SET citus.log_local_commands TO TRUE;`, adding that now

Comment (onderkalaci) [reply src/test/regress/expected/intermediate_result_pruning.out]: btw, In general, I found the number of tests that explicitly tests the edge cases of the PR is not here, so adding another test file with such queries

File: src/backend/distributed/executor/adaptive_executor.c [hunk 15]
```diff
@@ -1596,6 +1595,11 @@ AssignTasksToConnections(DistributedExecution *execution)
 		int placementExecutionIndex = 0;
 		int placementExecutionCount = list_length(task->taskPlacementList);
 
+		if (!TransactionConnectedToLocalhost)
+		{
+			TransactionConnectedToLocalhost = TaskAccessesLocalNode(task);
```
Comment (onderkalaci) [inline src/backend/distributed/executor/adaptive_executor.c]: I'll change this flag a bit more to support more cases such as . 
```SQL
	if (!TransactionConnectedToLocalhost)
		{
			TransactionConnectedToLocalhost = TaskAccessesLocalNode(task) && modLevel > ROW_MODIFY_READONLY;
		}
```

(I'll change the name)

File: src/backend/distributed/planner/multi_router_planner.c (hunk 7 above)
Comment (thanodnl) [reply src/backend/distributed/planner/multi_router_planner.c]: can this be reflected in the description of the PR?

File: src/backend/distributed/planner/intermediate_result_pruning.c (hunk 11 above)
Comment (thanodnl) [reply src/backend/distributed/planner/intermediate_result_pruning.c]: instead of having this complicated logic to remove the local node from the list can we rewrite the loop above (the one that sets `entry->writeLocalFile` and creates `workerNodeList` from 1 loop to 2 loops?

first loop checks for and sets `entry->writeLocalFile`
second loop constructs `workerNodeList`

if performance of running the loop twice is a concern we could also fold it in 1 loop and keep track of the `workerNode` we want to remove later during the run of the first loop (nothing to remove if ptr is `NULL`).

Nevertheless the current construct is controversial and hard to follow.

File: src/backend/distributed/planner/intermediate_result_pruning.c [hunk 16]
```diff
@@ -203,25 +211,66 @@ MakeIntermediateResultHTAB()
...
+				break;
+			}
+		}
+	}
+
+	/* now, log the summary */
```
Comment (thanodnl) [inline src/backend/distributed/planner/intermediate_result_pruning.c]: Lets expand this comment and explain when we log:
```C
/*
 * Log a summary of decisions made for local execution.
 * By default we log at level DEBUG4
 * When the user has set <name for guc that controls LogIntermediateResults> we change the log level to DEBUG1
 */
```

Also, we have this check here but always log at level `DEBUG1`. This cause the log lines to be prepended with `DEBUG1` even if we log them because we always log them on level `DEBUG4`. But now the user sees a `DEBUG1` message when level is set to `DEBUG4` and not when it is set to `DEBUG1`. This would be super confusing for a user that is debugging an issue.

Instead lets rewrite it to decide the log level separate from the check if the level is loggable:

```C
log_level = DEBUG4;
if (LogIntermediateResults) { log_level = DEBUG1; }
if (IsLoggableLevel(log_level)) {
    ....
    elog(log_level, ....);
    ....
}
```

**edit:** I now see this construct is not introduced in this PR, but I stand my ground around the difference in log level we check for and the log level we log on, huge no-no. So lets address it here anyway since we touch on the logging already in this PR.

File: src/include/distributed/adaptive_executor.h [hunk 17]
```diff
@@ -1,6 +1,10 @@
 #ifndef ADAPTIVE_EXECUTOR_H
 #define ADAPTIVE_EXECUTOR_H
 
+#include "postgres.h"
+#include "funcapi.h"
```
Comment (thanodnl) [inline src/include/distributed/adaptive_executor.h]: nothing should be included in the same block as `postgres.h` a lot of include files from postgres assume this header has already been included.

Once we add rules to uncrustify to reorder includes alphabetically this will break.
Instead it is better to include `funcapi.h` in the block below where we include both our and postgres' header files.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 18]
```diff
@@ -1589,13 +1588,20 @@ AssignTasksToConnections(DistributedExecution *execution)
...
 	foreach_ptr(task, taskList)
 	{
 		bool placementExecutionReady = true;
 		int placementExecutionIndex = 0;
 		int placementExecutionCount = list_length(task->taskPlacementList);
 
+		if (!TransactionConnectedToLocalhost)
+		{
+			/* connecting local host via remote connection, record this */
+			TransactionConnectedToLocalhost = TaskAccessesLocalNode(task);
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c]: makes sense

**Comment by JelteF** [comment]
@onderkalaci I guess this can be closed now that #3410 is merged?

**Comment by onderkalaci** [comment]
> @onderkalaci I guess this can be closed now that #3410 is merged?

Yes, thanks for reminding. This PR is implemented via #3410, the changes are similar.
